### DISCUSSION TITLE
Mocking Modules

### DISCUSSION THREAD

**Discussion by koddsson** [nodejs/help#4810] [Q&A] [answered]
Hey! I'm playing around with the new testing and mocking utilities and I'm running into a issue trying to mock a module. I've got something like this in my code:

```js
import assert from 'node:assert/strict';
import { test } from 'node:test';
import * as leftPad from 'left-pad';

import codeThatUsesLeftPad from '../index.mjs';

test('test code', async (t) => {
  t.mock.method(leftPad, 'foo', (...args) => console.log('foo: ', ...args));
  codeThatUsesLeftPad();
});
```

But I get this error:

```
TypeError: Cannot redefine property: foobar
      at defineProperty (<anonymous>)
      at MockTracker.method (node:internal/test_runner/mock:207:5)
```

Am I doing this wrong?

**Comment by cjihrig** [comment, marked as answer]
You haven't provided a runnable example, so it's hard to say for sure. However, given the error message, it's possible you're trying to mock a method that is [non-configurable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cant_redefine_property):

```js
import { test } from 'node:test';
const obj = Object.create(null);
Object.defineProperty(obj, 'foo', { value: function() {} });

test('test code', async (t) => {
  t.mock.method(obj, 'foo');
});
```

I can tell you that mocking of entire modules is not currently implemented. It's something I plan to work on once ESM loaders stabilize a bit more.

**Reply by koddsson** [reply]
> I can tell you that mocking of entire modules is not currently implemented. It's something I plan to work on once ESM loaders stabilize a bit more.

Ah, that answers my question. Thank you! I really like the test runner and mocking functions 🙏🏻

**Reply by jasonwilliams** [reply]
Hey @cjihrig thanks for your great work on this, is there an open issue for mocking modules? If so would it be worth someone making one to track it?

**Reply by cjihrig** [reply]
I don't think there is an open issue. Feel free to open one if you want. I do have a [work in progress branch](https://github.com/cjihrig/node/tree/mock-no-src). I think the biggest things missing before I can open a PR are related to ESM loaders and are tracked in https://github.com/nodejs/node/issues/49472 and https://github.com/nodejs/node/issues/49473.

**Reply by jasonwilliams** [reply]
Hey @cjihrig I made an issue here:
https://github.com/nodejs/node/issues/51164
